File: src/test/java/no/nav/fo/veilarbregistrering/kafka/FormidlingsgruppeMapperTest.java
package no.nav.fo.veilarbregistrering.kafka;

import no.nav.fo.veilarbregistrering.FileToJson;
import no.nav.fo.veilarbregistrering.arbeidssoker.Operation;
import no.nav.fo.veilarbregistrering.arbeidssoker.Formidlingsgruppe;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;

import static org.assertj.core.api.Assertions.assertThat;

public class FormidlingsgruppeMapperTest {

    @Test
    public void skal_mappe_json_fra_gg_arena_formidlingsgruppe_v1_til_formidlingsgruppeEvent() {
        String json = FileToJson.toJson("/kafka/formidlingsgruppe_uten_mod_dato.json");
        FormidlingsgruppeEvent formidlingsgruppeEvent = InsertAndUpdateFormidlingsgruppeMapper.map(json);

        assertThat(formidlingsgruppeEvent.getFoedselsnummer().get().stringValue()).isEqualTo("***********");
        assertThat(formidlingsgruppeEvent.getPersonId()).isEqualTo("13919");
        assertThat(formidlingsgruppeEvent.getFormidlingsgruppe()).isEqualTo(Formidlingsgruppe.of("ISERV"));
    }

    @Test
    public void skal_mappe_json_med_mod_dato_fra_gg_arena_formidlingsgruppe_v1_til_formidlingsgruppeEvent() {
        String json = FileToJson.toJson("/kafka/formidlingsgruppe_med_mod_dato.json");
        FormidlingsgruppeEvent formidlingsgruppeEvent = InsertAndUpdateFormidlingsgruppeMapper.map(json);

        assertThat(formidlingsgruppeEvent.getFoedselsnummer().get().stringValue()).isEqualTo("***********");
        assertThat(formidlingsgruppeEvent.getPersonId()).isEqualTo("3226568");
        assertThat(formidlingsgruppeEvent.getFormidlingsgruppe()).isEqualTo(Formidlingsgruppe.of("ARBS"));
        assertThat(formidlingsgruppeEvent.getFormidlingsgruppeEndret())
                .isEqualTo(LocalDateTime.of(2020, 6, 19,9,31,50));
    }

    @Test
    public void skal_mappe_json_uten_fnr_fra_gg_arena_formidlingsgruppe_v1_til_formidlingsgruppeEvent() {
        String json = FileToJson.toJson("/kafka/formidlingsgruppe_uten_fnr.json");
        FormidlingsgruppeEvent formidlingsgruppeEvent = InsertAndUpdateFormidlingsgruppeMapper.map(json);

        assertThat(formidlingsgruppeEvent.getFoedselsnummer()).isEmpty();
        assertThat(formidlingsgruppeEvent.getPersonId()).isEqualTo("1652");
        assertThat(formidlingsgruppeEvent.getOperation()).isEqualTo(Operation.INSERT);
        assertThat(formidlingsgruppeEvent.getFormidlingsgruppe()).isEqualTo(Formidlingsgruppe.of("ISERV"));
        assertThat(formidlingsgruppeEvent.getFormidlingsgruppeEndret())
                .isEqualTo(LocalDateTime.of(2007, 12, 3,3,5,54));
    }

    @Test
    public void skal_mappe_bade_after_og_before() {
        String json = FileToJson.toJson("/kafka/formidlingsgruppe_med_mod_dato.json");
        FormidlingsgruppeEvent formidlingsgruppeEvent = InsertAndUpdateFormidlingsgruppeMapper.map(json);

        assertThat(formidlingsgruppeEvent.getFoedselsnummer().get().stringValue()).isEqualTo("***********");
        assertThat(formidlingsgruppeEvent.getPersonId()).isEqualTo("3226568");
        assertThat(formidlingsgruppeEvent.getOperation()).isEqualTo(Operation.UPDATE);
        assertThat(formidlingsgruppeEvent.getFormidlingsgruppe()).isEqualTo(Formidlingsgruppe.of("ARBS"));
        assertThat(formidlingsgruppeEvent.getFormidlingsgruppeEndret())
                .isEqualTo(LocalDateTime.of(2020, 6, 19,9,31,50));
        assertThat(formidlingsgruppeEvent.getForrigeFormidlingsgruppe()).hasValue(Formidlingsgruppe.of("ISERV"));
        assertThat(formidlingsgruppeEvent.getForrigeFormidlingsgruppeEndret())
                .hasValue(LocalDateTime.of(2020, 6, 18,11,13,1));
    }

    @Test
    public void skal_mappe_json_med_opt_type_d_fra_gg_arena_formidlingsgruppe_v1_til_formidlingsgruppeEvent() {
        String json = FileToJson.toJson("/kafka/formidlingsgruppe_med_slett.json");
        FormidlingsgruppeEvent formidlingsgruppeEvent = InsertAndUpdateFormidlingsgruppeMapper.map(json);

        assertThat(formidlingsgruppeEvent.getFoedselsnummer()).isEmpty();
        assertThat(formidlingsgruppeEvent.getPersonId()).isEqualTo("1365747");
        assertThat(formidlingsgruppeEvent.getFormidlingsgruppe()).isEqualTo(Formidlingsgruppe.of("IJOBS"));
        assertThat(formidlingsgruppeEvent.getFormidlingsgruppeEndret())
                .isEqualTo(LocalDateTime.of(2016, 3, 12,0,47,50));
    }
}

File: src/test/kotlin/no/nav/fo/veilarbregistrering/arbeidssoker/formidlingsgruppe/kafka/FormidlingsgruppeMapperTest.kt
package no.nav.fo.veilarbregistrering.arbeidssoker.formidlingsgruppe.kafka

import no.nav.fo.veilarbregistrering.FileToJson.toJson
import no.nav.fo.veilarbregistrering.arbeidssoker.formidlingsgruppe.Formidlingsgruppe
import no.nav.fo.veilarbregistrering.arbeidssoker.formidlingsgruppe.Operation
import no.nav.fo.veilarbregistrering.arbeidssoker.formidlingsgruppe.kafka.FormidlingsgruppeMapper
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.time.LocalDateTime

class FormidlingsgruppeMapperTest {

    @Test
    fun `Kaster exception ved manglende mod dato`() {
        val json = toJson("/kafka/formidlingsgruppe_uten_mod_dato.json")
        assertThrows<IllegalArgumentException> { FormidlingsgruppeMapper.map(json) }
    }

    @Test
    fun `Kaster exception ved manglende op_type`() {
        val json = toJson("/kafka/formidlingsgruppe_uten_op_type_I.json")
        assertThrows<IllegalArgumentException> { FormidlingsgruppeMapper.map(json) }
    }

    @Test
    fun `Kaster exception ved manglende person_id_status`() {
        val json = toJson("/kafka/formidlingsgruppe_uten_person_id_status.json")
        assertThrows<IllegalArgumentException> { FormidlingsgruppeMapper.map(json) }
    }

    @Test
    fun `skal mappe json med mod dato fra gg arena formidlingsgruppe v1 til formidlingsgruppeEvent`() {
        val json = toJson("/kafka/formidlingsgruppe_med_mod_dato.json")
        val formidlingsgruppeEvent = FormidlingsgruppeMapper.map(json)
        assertThat(formidlingsgruppeEvent.foedselsnummer?.stringValue()).isEqualTo("***********")
        assertThat(formidlingsgruppeEvent.personId).isEqualTo("3226568")
        assertThat(formidlingsgruppeEvent.formidlingsgruppe).isEqualTo(Formidlingsgruppe("ARBS"))
        assertThat(formidlingsgruppeEvent.formidlingsgruppeEndret)
            .isEqualTo(LocalDateTime.of(2020, 6, 19, 9, 31, 50))
    }

    @Test
    fun `manglende fnr skal kaste exception`() {
        val json = toJson("/kafka/formidlingsgruppe_uten_fnr.json")
        assertThrows<IllegalArgumentException> { FormidlingsgruppeMapper.map(json) }
    }

    @Test
    fun `skal mappe både after og before`() {
        val json = toJson("/kafka/formidlingsgruppe_med_mod_dato.json")
        val formidlingsgruppeEvent = FormidlingsgruppeMapper.map(json)
        assertThat(formidlingsgruppeEvent.foedselsnummer?.stringValue()).isEqualTo("***********")
        assertThat(formidlingsgruppeEvent.personId).isEqualTo("3226568")
        assertThat(formidlingsgruppeEvent.operation).isEqualTo(Operation.UPDATE)
        assertThat(formidlingsgruppeEvent.formidlingsgruppe).isEqualTo(Formidlingsgruppe("ARBS"))
        assertThat(formidlingsgruppeEvent.formidlingsgruppeEndret)
            .isEqualTo(LocalDateTime.of(2020, 6, 19, 9, 31, 50))
        assertThat(formidlingsgruppeEvent.forrigeFormidlingsgruppe).isEqualTo(Formidlingsgruppe("ISERV"))
        assertThat(formidlingsgruppeEvent.forrigeFormidlingsgruppeEndret)
            .isEqualTo(LocalDateTime.of(2020, 6, 18, 11, 13, 1))
    }

    @Test
    fun `mapping av op type d for delete`() {
        val json = toJson("/kafka/formidlingsgruppe_op_type_D.json")
        val formidlingsgruppeEvent = FormidlingsgruppeMapper.map(json)
        assertThat(formidlingsgruppeEvent.foedselsnummer.stringValue()).isEqualTo("***********")
        assertThat(formidlingsgruppeEvent.personId).isEqualTo("1365747")
        assertThat(formidlingsgruppeEvent.operation).isEqualTo(Operation.DELETE)
        assertThat(formidlingsgruppeEvent.formidlingsgruppe).isEqualTo(Formidlingsgruppe("IJOBS"))
        assertThat(formidlingsgruppeEvent.formidlingsgruppeEndret)
            .isEqualTo(LocalDateTime.of(2016, 3, 12, 0, 47, 50))
    }
}
